### PR TITLE
Add Slack layer

### PR DESCRIPTION
**Pull request by kostajh**
This PR adds support for Slack via the the excellent [emacs-slack](https://github.com/yuya373/emacs-slack) package. n.b. that package is still in heavy development, but I and others are now using it as our primary clients for work, so IMO it's probably ready enough to be included in Spacemacs.

I based much of this layer off of the code in the ERC layer, and generally want to have a consistent experience across the IRC / Slack layers.

I could certainly use some help in bringing this to completion, so any suggestions or contributions are  welcome.

TODO:

- [ ] Add company-emoji integration
- [x] Fix flyspell integration
- [x] bind `@`, `#` and `Ctrl-K` in insert mode

### CODE REVIEW

**Comment by d12frosted** [comment]
:+1: Like the idea.

**Comment by StreakyCobra** [comment]
Wait… WHAT? that exists? Cool I will probably give it a try!

**Comment by robbyoconnor** [comment]
I love you so much right now! :heart:

**Comment by kostajh** [comment]
@robbyoconnor all credit is to @yuya373 :-)

**Comment by robbyoconnor** [comment]
@yuya373 :heart:

**Comment by kostajh** [comment]
The only thing missing is the `company-emoji` integration, if anyone wants to add it. Otherwise I think this is more or less ready.

**Comment by kostajh** [comment]
Ok, well one more thing. Binding `@` (user mention embed), `#` (channel mention embed) and `Ctrl-k` (switch channel) during insert mode would be very handy.

**Comment by tadfisher** [comment]
Why does this include changes to the `+lang/php` layer?

**Comment by robbyoconnor** [comment]
Probably a rebase issue.

--Rob

**Comment by kostajh** [comment]
Yeah, my fault, I'll fix.


On Sat, Feb 6, 2016, at 06:12 PM, robbyoconnor wrote:
> Probably a rebase issue.
>
>
--Rob
>
Sent from my cell, please excuse any typos.
>
On Feb 6, 2016 6:11 PM, "Tad Fisher" <notifications@github.com> wrote:
>
>
> Why does this include changes to the +lang/php layer?
>
>
>
> —
>
> Reply to this email directly or view it on GitHub
>
> <https://github.com/syl20bnr/spacemacs/pull/4821#issuecomment-180882733>
> .
>
>
> — Reply to this email directly or view it on GitHub[1].


>



Links:

  1. https://github.com/syl20bnr/spacemacs/pull/4821#issuecomment-180882996

**Comment by robbyoconnor** [comment]
@kostajh what the heck happened? You should be rebasing `git pull --rebase` not merging...

**Comment by kostajh** [comment]
I've been working on the PHP layer in another PR. I have just been
rebasing but must have missed something.


On Sat, Feb 6, 2016, at 09:10 PM, robbyoconnor wrote:
> @kostajh[1] what the heck happened? You should be rebasing git pull
> --rebase not merging...


> — Reply to this email directly or view it on GitHub[2].


>



Links:

  1. https://github.com/kostajh
  2. https://github.com/syl20bnr/spacemacs/pull/4821#issuecomment-180913317

**Comment by robbyoconnor** [comment]
Each branch should ONLY rebase `upstream/develop` -- never your local (`origin`). So in theory what happened here should never happen.

**Comment by kostajh** [comment]
Right. Don't worry it's not the end of the world! It will get fixed.


On Sat, Feb 6, 2016, at 09:50 PM, robbyoconnor wrote:
> Each branch should ONLY rebase upstream/develop -- never your local
> (origin). So in theory what happened here should never happen.


> — Reply to this email directly or view it on GitHub[1].


>



Links:

  1. https://github.com/syl20bnr/spacemacs/pull/4821#issuecomment-180924508

**Comment by robbyoconnor** [comment]
it is totally the end of the world! I just saw the news, you caused the end of the world! </troll>

**Comment by kostajh** [comment]
Everyone happy now? :-P I may have sinned in [this line](https://github.com/syl20bnr/spacemacs/pull/4821/files#diff-39a46ee409e1f24f67f7d904dd6312b3R76) so if there's a more correct way to implement this, let me know.

File: layers/slack/packages.el [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+      "mm" 'slack-message-embed-mention
+      "mc" 'slack-message-embed-channel
+      "k" 'slack-channel-select
+      "@" 'slack-message-embed-mention
+      "#" 'slack-message-embed-channel
+      )
+    (evil-define-key 'insert slack-mode-map
+      (kbd "C-k") 'slack-channel-select
+      (kbd "@") 'slack-message-embed-mention
+      (kbd "#") 'slack-message-embed-channel)
```
Comment (nixmaniack) [inline layers/slack/packages.el]: I think `@` and `#` are fine here. But use of `C-k` in `insert` state looks undesirable. (I don't of proper way to suggest something here)

Comment (StreakyCobra) [inline layers/slack/packages.el]: Indeed, `C-k` is used by people who don't have accents on their keyboard layout to create accented characters (`C-k ' e` to create `é` for instance). This is especially useful in slack as I suppose you may be discussing in your mother tongue.

Comment (kostajh) [inline layers/slack/packages.el]: Makes sense, I'll remove `C-k`


On Mon, Feb 8, 2016, at 07:30 AM, Fabien Dubosson wrote:
> In layers/slack/packages.el[1]:


> > +    (spacemacs/set-leader-keys-for-major-mode 'slack-mode
> +      "j" 'slack-channel-select
> +      "d" 'slack-im-select
> +      "p" 'slack-room-load-prev-messages
> +      "e" 'slack-message-edit
> +      "q" 'slack-ws-close
> +      "mm" 'slack-message-embed-mention
> +      "mc" 'slack-message-embed-channel
> +      "k" 'slack-channel-select
> +      "@" 'slack-message-embed-mention
> +      "#" 'slack-message-embed-channel
> +      )
> +    (evil-define-key 'insert slack-mode-map
> +      (kbd "C-k") 'slack-channel-select
> +      (kbd "@") 'slack-message-embed-mention
> +      (kbd "#") 'slack-message-embed-channel) Indeed, C-k is used by
>        people who don't have accents on their keyboard layout to
>        create accented characters (C-k ' e to create é for instance).
>        This is especially useful in slack as I suppose you may be
>        discussing in your mother tongue.


> — Reply to this email directly or view it on GitHub[2].


>



Links:

  1. https://github.com/syl20bnr/spacemacs/pull/4821#discussion_r52162739
  2. https://github.com/syl20bnr/spacemacs/pull/4821/files#r52162739

**Comment by ericfode** [comment]
Tested this out, works wonderfully :+1:  Thanks for packaging this up! @kostajh

**Comment by a13ph** [comment]
BTW there are ways to use Slack through IRC or XMPP
https://get.slack.help/hc/en-us/articles/201727913-Connecting-to-Slack-over-IRC-and-XMPP

**Comment by tadfisher** [comment]
@a13ph And both of those ways are terrible if your teammates use any Slack features.

**Comment by a13ph** [comment]
@tadfisher which are most incompatible? 

ERC for comparison does have some features mimicking Gitter ones. I assume this approach would work for most Slack of stuff too.

**Comment by tadfisher** [comment]
The Slack IRC and XMPP gateways themselves are the limiting factor here. Things like:

- Editing messages
- Deleting messages
- Emoji
- Private channels
- Code blocks
- Markdown syntax
- Media attachments
- Snippets
- File uploads

Are either straight-up broken or produce strange, garbagey output in an IRC client, no matter if it's ERC or something else.

**Comment by a13ph** [comment]
I can see potential issues only with implementing private channels without upstream support, tbh. Everything else is feasible, even if some things are harder than others. But if gateway is an issue - it's not frontend matter, must bring it to the upstream

**Comment by tadfisher** [comment]
Why should ERC implement features to work around Slack's IRC gateway problems? ERC is an IRC client, not a Slack IRC Gateway client.

**Comment by kostajh** [comment]
Just seconding the view that Slack IRC is a terrible way to use the service. One thing not mentioned above: the ability to view the channel's history/context when joining it is invaluable, and something not possible with the IRC gateway unless you put a fair amount of time and effort into configuring ZNC or using a cloud IRC service.

**Comment by sideshowdave7** [comment]
Just a note this is broken with latest of emacs-slack as they got rid of:

`(defcustom slack-client-id)`, etc to support multiple slack configurations.

You'll need to run something like this instead:

`(slack-register-team`
  ` :name "emacs-slack"`
 `  :default t`
 `  :client-id "aaaaaaaaaaa.00000000000"`
   `:client-secret "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"`
   `:token "aaaa-sssssssssss-88888888888-hhhhhhhhhhh-jjjjjjjjjj"`
   `:subscribed-channels '(test-rename rrrrr))`

**Comment by robbyoconnor** [comment]
Fork it and make a PR @sideshowdave7 -- unless @kostajh  plans to fix it

**Comment by kostajh** [comment]
Sure, I can update the PR. I'm using this layer without issue. I'll work on updating the docs this week.

**Comment by kostajh** [comment]
The README is updated thanks to @sideshowdave7. IMO the company-emoji bit could be added later. Otherwise this PR is ready.

**Comment by robbyoconnor** [comment]
Hi,

I'd like to add you to my professional network on LinkedIn.

- Robert

Confirm that you know Robert: https://www.linkedin.com/e/v2?e=0-ini67ohs-jy&t=ssuw&tracking=eml-guest-invite-cta&ek=invite_guest&sharedKey=fkR9rdtu&invitationId=6130896990847979520








You received an invitation to connect. LinkedIn will use your email address to make suggestions to our members in features like People You May Know. Unsubscribe here: https://www.linkedin.com/e/v2?e=0-ini67ohs-jy&t=lun&midToken=AQFmrFVhGEd36Q&ek=invite_guest&loid=AQHIH5PUuw5zwwAAAVRVScVp0BzVoDk49CU9jtvKv_w2BEODmNOlyBlNmFCa0532M21Y6Y9JFFxl9t-RWPcIoSoxFlcmCddNEKc6T8teb1_g7etF-9Y7VvNv8pHef9uTZRjI_5fAivaLUsaRZ3AbAqa0U1cuweXzKtHfrmOlGg8wWbyjw63s0QgRx7uh2GnkG9hl&eid=0-ini67ohs-jy


This email was sent to reply@reply.github.com.

If you need assistance or have questions, please contact LinkedIn Customer Service: https://www.linkedin.com/e/v2?e=0-ini67ohs-jy&a=customerServiceUrl&ek=invite_guest

&copy; 2016 LinkedIn Corporation, 2029 Stierlin Court, Mountain View CA 94043. LinkedIn and the LinkedIn logo are registered trademarks of LinkedIn.

**Comment by kostajh** [comment]
Hmm. Well it's not quite ready any more, as I see this when I try to switch to a channel:

**Comment by d12frosted** [comment]
This might be not related. See [this](https://github.com/syl20bnr/spacemacs/search?q=spacemacs%2F%2Funset-scroll-margin&type=Issues&utf8=✓).

**Comment by kostajh** [comment]
Ah, thanks for the heads up. That's surely the same issue.

**Comment by syl20bnr** [comment]
Thank you ! 👍 
I made some tweaks and fixes the hook issue in commit https://github.com/syl20bnr/spacemacs/commit/2262bbc
Cherry-picked into develop branch, you can safely delete your branch.

**Pull request closed without merging**